api/osb/reference_instance_plugin.go

```diff
@@ -0,0 +1,363 @@
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/pkg/web"
+	"github.com/Peripli/service-manager/storage"
+	"github.com/tidwall/gjson"
+	"net/http"
+)
+
+const ReferenceInstancePluginName = "ReferenceInstancePlugin"
+const planIDProperty = "plan_id"
+const catalogIDProperty = "catalog_id"
+const contextKey = "context"
+
+type referenceInstancePlugin struct {
+	repository       storage.TransactionalRepository
+	tenantIdentifier string
+}
+
+type osbObject = map[string]interface{}
+
+// NewCheckPlatformIDPlugin creates new plugin that checks the platform_id of the instance
+func NewReferenceInstancePlugin(repository storage.TransactionalRepository, tenantIdentifier string) *referenceInstancePlugin {
+	return &referenceInstancePlugin{
+		repository:       repository,
+		tenantIdentifier: tenantIdentifier,
+	}
+}
+
+// Name returns the name of the plugin
+func (p *referenceInstancePlugin) Name() string {
+	return ReferenceInstancePluginName
+}
+
+func (p *referenceInstancePlugin) Provision(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	servicePlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	isReferencePlan, err := p.isReferencePlan(ctx, catalogIDProperty, servicePlanID)
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	// Ownership validation
+	callerTenantID := gjson.GetBytes(req.Body, "context."+p.tenantIdentifier).String()
+	if callerTenantID != "" {
+		err = p.validateOwnership(req)
+		if err != nil {
+			return nil, err
+		}
+	}
+	parameters := gjson.GetBytes(req.Body, "parameters").Map()
+	referencedInstanceID, exists := parameters[constant.ReferencedInstanceIDKey]
+	if !exists {
+		return nil, util.HandleInstanceSharingError(util.ErrMissingReferenceParameter, constant.ReferencedInstanceIDKey)
+	}
+	_, err = p.isReferencedShared(ctx, referencedInstanceID.String())
+	if err != nil {
+		return nil, err
+	}
+	return p.generateOSBResponse(ctx, "Provision", nil)
+}
+
+// Deprovision intercepts deprovision requests and check if the instance is in the platform from where the request comes
+func (p *referenceInstancePlugin) Deprovision(req *web.Request, next web.Handler) (*web.Response, error) {
+	instanceID := req.PathParams["instance_id"]
+	if instanceID == "" {
+		return next.Handle(req)
+	}
+	ctx := req.Context()
+
+	dbInstanceObject, err := p.getObjectByOperator(ctx, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+
+	isReferencePlan, err := p.isReferencePlan(ctx, "id", instance.ServicePlanID)
+
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	return p.generateOSBResponse(ctx, "Deprovision", nil)
```

Comment:
consider defining all possible types for this function.
maybe something like:
type OperationCategory string

const (
	// CREATE represents an operation type for creating a resource
	CREATE OperationCategory = "create"

	// UPDATE represents an operation type for updating a resource
	UPDATE OperationCategory = "update"

	// DELETE represents an operation type for deleting a resource
	DELETE OperationCategory = "delete"
)